plugins/hls-eval-plugin/src/Ide/Plugin/Eval/Rules.hs

```diff
@@ -57,12 +58,9 @@ queueForEvaluation ide nfp = do
     modifyIORef var (Set.insert nfp)
 
 #if MIN_VERSION_ghc(9,2,0)
-getAnnotations :: Development.IDE.GHC.Compat.Located HsModule -> [LEpaComment]
-getAnnotations (L _ (HsModule { hsmodAnn = anns'})) = priorComments $ epAnnComments anns'
-
 apiAnnComments' :: ParsedModule -> [SrcLoc.RealLocated EpaCommentTok]
 apiAnnComments' pm = do
-  L span' (EpaComment c span) <- getAnnotations $ pm_parsed_source pm
+  L span' (EpaComment c span) <- listify (const True) $ pm_parsed_source pm
```

Comment:
Beautiful but very inefficient. Isn't it enough with looking at the `priorComments` and the `followingComments` of the `hsModAnn` ?

Reply:
Unfortunately, `followingComments` does not do the trick.

I'll fix the unit test first, and then works on a more efficient solution. I'm lost in the middle of `ghc` documentation in order to understand where are located the comment I care about. I would love if `syb` was able to give me the comment as well as a "path" where it was found.

Reply:
I used `gshow` on the AST for a simple example:

```haskell
module T1 where

import Data.List (unwords)

-- >>> unwords example
example :: [String]
example = ["This","is","an","example","of","evaluation"]

-- >>> 1 + 1
-- 2
```

And with a bit of reformatting, I get:

<details>
```haskell
f =
  ( L
      ( "{abstract:SrcSpan}"
      )
      ( HsModule
          ( EpAnn
              ( Anchor
                  ( "{abstract:RealSrcSpan}"
                  )
                  ( UnchangedAnchor
                  )
              )
              ( AnnsModule
                  ( ( :
                    )
                    ( AddEpAnn
                        ( AnnModule
                        )
                        ( EpaSpan
                            ( "{abstract:RealSrcSpan}"
                            )
                        )
                    )
                    ( ( :
                      )
                      ( AddEpAnn
                          ( AnnWhere
                          )
                          ( EpaSpan
                              ( "{abstract:RealSrcSpan}"
                              )
                          )
                      )
                      ( []
                      )
                    )
                  )
                  ( AnnList
                      ( Nothing
                      )
                      ( Nothing
                      )
                      ( Nothing
                      )
                      ( []
                      )
                      ( []
                      )
                  )
              )
              ( EpaCommentsBalanced
                  ( []
                  )
                  ( ( :
                    )
                    ( L
                        ( Anchor
                            ( "{abstract:RealSrcSpan}"
                            )
                            ( UnchangedAnchor
                            )
                        )
                        ( EpaComment
                            ( EpaEofComment
                            )
                            ( "{abstract:RealSrcSpan}"
                            )
                        )
                    )
                    ( []
                    )
                  )
              )
          )
          ( VirtualBraces
              ( 1
              )
          )
          ( Just
              ( L
                  ( SrcSpanAnn
                      ( EpAnnNotUsed
                      )
                      ( "{abstract:SrcSpan}"
                      )
                  )
                  ( "{abstract:ModuleName}"
                  )
              )
          )
          ( Nothing
          )
          ( ( :
            )
            ( L
                ( SrcSpanAnn
                    ( EpAnn
                        ( Anchor
                            ( "{abstract:RealSrcSpan}"
                            )
                            ( UnchangedAnchor
                            )
                        )
                        ( AnnListItem
                            ( []
                            )
                        )
                        ( EpaComments
                            ( ( :
                              )
                              ( L
                                  ( Anchor
                                      ( "{abstract:RealSrcSpan}"
                                      )
                                      ( UnchangedAnchor
                                      )
                                  )
                                  ( EpaComment
                                      ( EpaLineComment "-- >>> unwords example"
                                      )
                                      ( "{abstract:RealSrcSpan}"
                                      )
                                  )
                              )
                              ( []
                              )
                            )
                        )
                    )
                    ( "{abstract:SrcSpan}"
                    )
                )
                ( ImportDecl
                    ( EpAnn
                        ( Anchor
                            ( "{abstract:RealSrcSpan}"
                            )
                            ( UnchangedAnchor
                            )
                        )
                        ( EpAnnImportDecl
                            ( EpaSpan
                                ( "{abstract:RealSrcSpan}"
                                )
                            )
                            ( Nothing
                            )
                            ( Nothing
                            )
                            ( Nothing
                            )
                            ( Nothing
                            )
                            ( Nothing
                            )
                        )
                        ( EpaComments
                            ( []
                            )
                        )
                    )
                    ( NoSourceText
                    )
                    ( L
                        ( SrcSpanAnn
                            ( EpAnnNotUsed
                            )
                            ( "{abstract:SrcSpan}"
                            )
                        )
                        ( "{abstract:ModuleName}"
                        )
                    )
                    ( Nothing
                    )
                    ( NotBoot
                    )
                    ( False
                    )
                    ( NotQualified
                    )
                    ( False
                    )
                    ( Nothing
                    )
                    ( Just
                        ( (,)
                          ( False
                          )
                          ( L
                              ( SrcSpanAnn
                                  ( EpAnn
                                      ( Anchor
                                          ( "{abstract:RealSrcSpan}"
                                          )
                                          ( UnchangedAnchor
                                          )
                                      )
                                      ( AnnList
                                          ( Nothing
                                          )
                                          ( Just
                                              ( AddEpAnn
                                                  ( AnnOpenP
                                                  )
                                                  ( EpaSpan
                                                      ( "{abstract:RealSrcSpan}"
                                                      )
                                                  )
                                              )
                                          )
                                          ( Just
                                              ( AddEpAnn
                                                  ( AnnCloseP
                                                  )
                                                  ( EpaSpan
                                                      ( "{abstract:RealSrcSpan}"
                                                      )
                                                  )
                                              )
                                          )
                                          ( []
                                          )
                                          ( []
                                          )
                                      )
                                      ( EpaComments
                                          ( []
                                          )
                                      )
                                  )
                                  ( "{abstract:SrcSpan}"
                                  )
                              )
                              ( ( :
                                )
                                ( L
                                    ( SrcSpanAnn
                                        ( EpAnnNotUsed
                                        )
                                        ( "{abstract:SrcSpan}"
                                        )
                                    )
                                    ( IEVar
                                        ( NoExtField
                                        )
                                        ( L
                                            ( SrcSpanAnn
                                                ( EpAnnNotUsed
                                                )
                                                ( "{abstract:SrcSpan}"
                                                )
                                            )
                                            ( IEName
                                                ( L
                                                    ( SrcSpanAnn
                                                        ( EpAnnNotUsed
                                                        )
                                                        ( "{abstract:SrcSpan}"
                                                        )
                                                    )
                                                    ( Unqual
                                                        ( "{abstract:OccName}"
                                                        )
                                                    )
                                                )
                                            )
                                        )
                                    )
                                )
                                ( []
                                )
                              )
                          )
                        )
                    )
                )
            )
            ( []
            )
          )
          ( ( :
            )
            ( L
                ( SrcSpanAnn
                    ( EpAnn
                        ( Anchor
                            ( "{abstract:RealSrcSpan}"
                            )
                            ( UnchangedAnchor
                            )
                        )
                        ( AnnListItem
                            ( []
                            )
                        )
                        ( EpaComments
                            ( []
                            )
                        )
                    )
                    ( "{abstract:SrcSpan}"
                    )
                )
                ( SigD
                    ( NoExtField
                    )
                    ( TypeSig
                        ( EpAnn
                            ( Anchor
                                ( "{abstract:RealSrcSpan}"
                                )
                                ( UnchangedAnchor
                                )
                            )
                            ( AnnSig
                                ( AddEpAnn
                                    ( AnnDcolon
                                    )
                                    ( EpaSpan
                                        ( "{abstract:RealSrcSpan}"
                                        )
                                    )
                                )
                                ( []
                                )
                            )
                            ( EpaComments
                                ( []
                                )
                            )
                        )
                        ( ( :
                          )
                          ( L
                              ( SrcSpanAnn
                                  ( EpAnnNotUsed
                                  )
                                  ( "{abstract:SrcSpan}"
                                  )
                              )
                              ( Unqual
                                  ( "{abstract:OccName}"
                                  )
                              )
                          )
                          ( []
                          )
                        )
                        ( HsWC
                            ( NoExtField
                            )
                            ( L
                                ( SrcSpanAnn
                                    ( EpAnnNotUsed
                                    )
                                    ( "{abstract:SrcSpan}"
                                    )
                                )
                                ( HsSig
                                    ( NoExtField
                                    )
                                    ( HsOuterImplicit
                                        ( NoExtField
                                        )
                                    )
                                    ( L
                                        ( SrcSpanAnn
                                            ( EpAnnNotUsed
                                            )
                                            ( "{abstract:SrcSpan}"
                                            )
                                        )
                                        ( HsListTy
                                            ( EpAnn
                                                ( Anchor
                                                    ( "{abstract:RealSrcSpan}"
                                                    )
                                                    ( UnchangedAnchor
                                                    )
                                                )
                                                ( AnnParen
                                                    ( AnnParensSquare
                                                    )
                                                    ( EpaSpan
                                                        ( "{abstract:RealSrcSpan}"
                                                        )
                                                    )
                                                    ( EpaSpan
                                                        ( "{abstract:RealSrcSpan}"
                                                        )
                                                    )
                                                )
                                                ( EpaComments
                                                    ( []
                                                    )
                                                )
                                            )
                                            ( L
                                                ( SrcSpanAnn
                                                    ( EpAnnNotUsed
                                                    )
                                                    ( "{abstract:SrcSpan}"
                                                    )
                                                )
                                                ( HsTyVar
                                                    ( EpAnn
                                                        ( Anchor
                                                            ( "{abstract:RealSrcSpan}"
                                                            )
                                                            ( UnchangedAnchor
                                                            )
                                                        )
                                                        ( []
                                                        )
                                                        ( EpaComments
                                                            ( []
                                                            )
                                                        )
                                                    )
                                                    ( NotPromoted
                                                    )
                                                    ( L
                                                        ( SrcSpanAnn
                                                            ( EpAnnNotUsed
                                                            )
                                                            ( "{abstract:SrcSpan}"
                                                            )
                                                        )
                                                        ( Unqual
                                                            ( "{abstract:OccName}"
                                                            )
                                                        )
                                                    )
                                                )
                                            )
                                        )
                                    )
                                )
                            )
                        )
                    )
                )
            )
            ( ( :
              )
              ( L
                  ( SrcSpanAnn
                      ( EpAnn
                          ( Anchor
                              ( "{abstract:RealSrcSpan}"
                              )
                              ( UnchangedAnchor
                              )
                          )
                          ( AnnListItem
                              ( []
                              )
                          )
                          ( EpaComments
                              ( ( :
                                )
                                ( L
                                    ( Anchor
                                        ( "{abstract:RealSrcSpan}"
                                        )
                                        ( UnchangedAnchor
                                        )
                                    )
                                    ( EpaComment
                                        ( EpaLineComment "-- 2"
                                        )
                                        ( "{abstract:RealSrcSpan}"
                                        )
                                    )
                                )
                                ( ( :
                                  )
                                  ( L
                                      ( Anchor
                                          ( "{abstract:RealSrcSpan}"
                                          )
                                          ( UnchangedAnchor
                                          )
                                      )
                                      ( EpaComment
                                          ( EpaLineComment "-- >>> 1 + 1"
                                          )
                                          ( "{abstract:RealSrcSpan}"
                                          )
                                      )
                                  )
                                  ( []
                                  )
                                )
                              )
                          )
                      )
                      ( "{abstract:SrcSpan}"
                      )
                  )
                  ( ValD
                      ( NoExtField
                      )
                      ( FunBind
                          ( NoExtField
                          )
                          ( L
                              ( SrcSpanAnn
                                  ( EpAnnNotUsed
                                  )
                                  ( "{abstract:SrcSpan}"
                                  )
                              )
                              ( Unqual
                                  ( "{abstract:OccName}"
                                  )
                              )
                          )
                          ( MG
                              ( NoExtField
                              )
                              ( L
                                  ( SrcSpanAnn
                                      ( EpAnnNotUsed
                                      )
                                      ( "{abstract:SrcSpan}"
                                      )
                                  )
                                  ( ( :
                                    )
                                    ( L
                                        ( SrcSpanAnn
                                            ( EpAnnNotUsed
                                            )
                                            ( "{abstract:SrcSpan}"
                                            )
                                        )
                                        ( Match
                                            ( EpAnn
                                                ( Anchor
                                                    ( "{abstract:RealSrcSpan}"
                                                    )
                                                    ( UnchangedAnchor
                                                    )
                                                )
                                                ( []
                                                )
                                                ( EpaComments
                                                    ( []
                                                    )
                                                )
                                            )
                                            ( FunRhs
                                                ( L
                                                    ( SrcSpanAnn
                                                        ( EpAnnNotUsed
                                                        )
                                                        ( "{abstract:SrcSpan}"
                                                        )
                                                    )
                                                    ( Unqual
                                                        ( "{abstract:OccName}"
                                                        )
                                                    )
                                                )
                                                ( Prefix
                                                )
                                                ( NoSrcStrict
                                                )
                                            )
                                            ( []
                                            )
                                            ( GRHSs
                                                ( EpaComments
                                                    ( []
                                                    )
                                                )
                                                ( ( :
                                                  )
                                                  ( L
                                                      ( "{abstract:SrcSpan}"
                                                      )
                                                      ( GRHS
                                                          ( EpAnn
                                                              ( Anchor
                                                                  ( "{abstract:RealSrcSpan}"
                                                                  )
                                                                  ( UnchangedAnchor
                                                                  )
                                                              )
                                                              ( GrhsAnn
                                                                  ( Nothing
                                                                  )
                                                                  ( AddEpAnn
                                                                      ( AnnEqual
                                                                      )
                                                                      ( EpaSpan
                                                                          ( "{abstract:RealSrcSpan}"
                                                                          )
                                                                      )
                                                                  )
                                                              )
                                                              ( EpaComments
                                                                  ( []
                                                                  )
                                                              )
                                                          )
                                                          ( []
                                                          )
                                                          ( L
                                                              ( SrcSpanAnn
                                                                  ( EpAnnNotUsed
                                                                  )
                                                                  ( "{abstract:SrcSpan}"
                                                                  )
                                                              )
                                                              ( ExplicitList
                                                                  ( EpAnn
                                                                      ( Anchor
                                                                          ( "{abstract:RealSrcSpan}"
                                                                          )
                                                                          ( UnchangedAnchor
                                                                          )
                                                                      )
                                                                      ( AnnList
                                                                          ( Nothing
                                                                          )
                                                                          ( Just
                                                                              ( AddEpAnn
                                                                                  ( AnnOpenS
                                                                                  )
                                                                                  ( EpaSpan
                                                                                      ( "{abstract:RealSrcSpan}"
                                                                                      )
                                                                                  )
                                                                              )
                                                                          )
                                                                          ( Just
                                                                              ( AddEpAnn
                                                                                  ( AnnCloseS
                                                                                  )
                                                                                  ( EpaSpan
                                                                                      ( "{abstract:RealSrcSpan}"
                                                                                      )
                                                                                  )
                                                                              )
                                                                          )
                                                                          ( []
                                                                          )
                                                                          ( []
                                                                          )
                                                                      )
                                                                      ( EpaComments
                                                                          ( []
                                                                          )
                                                                      )
                                                                  )
                                                                  ( ( :
                                                                    )
                                                                    ( L
                                                                        ( SrcSpanAnn
                                                                            ( EpAnn
                                                                                ( Anchor
                                                                                    ( "{abstract:RealSrcSpan}"
                                                                                    )
                                                                                    ( UnchangedAnchor
                                                                                    )
                                                                                )
                                                                                ( AnnListItem
                                                                                    ( ( :
                                                                                      )
                                                                                      ( AddCommaAnn
                                                                                          ( EpaSpan
                                                                                              ( "{abstract:RealSrcSpan}"
                                                                                              )
                                                                                          )
                                                                                      )
                                                                                      ( []
                                                                                      )
                                                                                    )
                                                                                )
                                                                                ( EpaComments
                                                                                    ( []
                                                                                    )
                                                                                )
                                                                            )
                                                                            ( "{abstract:SrcSpan}"
                                                                            )
                                                                        )
                                                                        ( HsLit
                                                                            ( EpAnn
                                                                                ( Anchor
                                                                                    ( "{abstract:RealSrcSpan}"
                                                                                    )
                                                                                    ( UnchangedAnchor
                                                                                    )
                                                                                )
                                                                                ( NoEpAnns
                                                                                )
                                                                                ( EpaComments
                                                                                    ( []
                                                                                    )
                                                                                )
                                                                            )
                                                                            ( HsString
                                                                                ( SourceText "\"This\""
                                                                                )
                                                                                ( "{abstract:FastString}"
                                                                                )
                                                                            )
                                                                        )
                                                                    )
                                                                    ( ( :
                                                                      )
                                                                      ( L
                                                                          ( SrcSpanAnn
                                                                              ( EpAnn
                                                                                  ( Anchor
                                                                                      ( "{abstract:RealSrcSpan}"
                                                                                      )
                                                                                      ( UnchangedAnchor
                                                                                      )
                                                                                  )
                                                                                  ( AnnListItem
                                                                                      ( ( :
                                                                                        )
                                                                                        ( AddCommaAnn
                                                                                            ( EpaSpan
                                                                                                ( "{abstract:RealSrcSpan}"
                                                                                                )
                                                                                            )
                                                                                        )
                                                                                        ( []
                                                                                        )
                                                                                      )
                                                                                  )
                                                                                  ( EpaComments
                                                                                      ( []
                                                                                      )
                                                                                  )
                                                                              )
                                                                              ( "{abstract:SrcSpan}"
                                                                              )
                                                                          )
                                                                          ( HsLit
                                                                              ( EpAnn
                                                                                  ( Anchor
                                                                                      ( "{abstract:RealSrcSpan}"
                                                                                      )
                                                                                      ( UnchangedAnchor
                                                                                      )
                                                                                  )
                                                                                  ( NoEpAnns
                                                                                  )
                                                                                  ( EpaComments
                                                                                      ( []
                                                                                      )
                                                                                  )
                                                                              )
                                                                              ( HsString
                                                                                  ( SourceText "\"is\""
                                                                                  )
                                                                                  ( "{abstract:FastString}"
                                                                                  )
                                                                              )
                                                                          )
                                                                      )
                                                                      ( ( :
                                                                        )
                                                                        ( L
                                                                            ( SrcSpanAnn
                                                                                ( EpAnn
                                                                                    ( Anchor
                                                                                        ( "{abstract:RealSrcSpan}"
                                                                                        )
                                                                                        ( UnchangedAnchor
                                                                                        )
                                                                                    )
                                                                                    ( AnnListItem
                                                                                        ( ( :
                                                                                          )
                                                                                          ( AddCommaAnn
                                                                                              ( EpaSpan
                                                                                                  ( "{abstract:RealSrcSpan}"
                                                                                                  )
                                                                                              )
                                                                                          )
                                                                                          ( []
                                                                                          )
                                                                                        )
                                                                                    )
                                                                                    ( EpaComments
                                                                                        ( []
                                                                                        )
                                                                                    )
                                                                                )
                                                                                ( "{abstract:SrcSpan}"
                                                                                )
                                                                            )
                                                                            ( HsLit
                                                                                ( EpAnn
                                                                                    ( Anchor
                                                                                        ( "{abstract:RealSrcSpan}"
                                                                                        )
                                                                                        ( UnchangedAnchor
                                                                                        )
                                                                                    )
                                                                                    ( NoEpAnns
                                                                                    )
                                                                                    ( EpaComments
                                                                                        ( []
                                                                                        )
                                                                                    )
                                                                                )
                                                                                ( HsString
                                                                                    ( SourceText "\"an\""
                                                                                    )
                                                                                    ( "{abstract:FastString}"
                                                                                    )
                                                                                )
                                                                            )
                                                                        )
                                                                        ( ( :
                                                                          )
                                                                          ( L
                                                                              ( SrcSpanAnn
                                                                                  ( EpAnn
                                                                                      ( Anchor
                                                                                          ( "{abstract:RealSrcSpan}"
                                                                                          )
                                                                                          ( UnchangedAnchor
                                                                                          )
                                                                                      )
                                                                                      ( AnnListItem
                                                                                          ( ( :
                                                                                            )
                                                                                            ( AddCommaAnn
                                                                                                ( EpaSpan
                                                                                                    ( "{abstract:RealSrcSpan}"
                                                                                                    )
                                                                                                )
                                                                                            )
                                                                                            ( []
                                                                                            )
                                                                                          )
                                                                                      )
                                                                                      ( EpaComments
                                                                                          ( []
                                                                                          )
                                                                                      )
                                                                                  )
                                                                                  ( "{abstract:SrcSpan}"
                                                                                  )
                                                                              )
                                                                              ( HsLit
                                                                                  ( EpAnn
                                                                                      ( Anchor
                                                                                          ( "{abstract:RealSrcSpan}"
                                                                                          )
                                                                                          ( UnchangedAnchor
                                                                                          )
                                                                                      )
                                                                                      ( NoEpAnns
                                                                                      )
                                                                                      ( EpaComments
                                                                                          ( []
                                                                                          )
                                                                                      )
                                                                                  )
                                                                                  ( HsString
                                                                                      ( SourceText "\"example\""
                                                                                      )
                                                                                      ( "{abstract:FastString}"
                                                                                      )
                                                                                  )
                                                                              )
                                                                          )
                                                                          ( ( :
                                                                            )
                                                                            ( L
                                                                                ( SrcSpanAnn
                                                                                    ( EpAnn
                                                                                        ( Anchor
                                                                                            ( "{abstract:RealSrcSpan}"
                                                                                            )
                                                                                            ( UnchangedAnchor
                                                                                            )
                                                                                        )
                                                                                        ( AnnListItem
                                                                                            ( ( :
                                                                                              )
                                                                                              ( AddCommaAnn
                                                                                                  ( EpaSpan
                                                                                                      ( "{abstract:RealSrcSpan}"
                                                                                                      )
                                                                                                  )
                                                                                              )
                                                                                              ( []
                                                                                              )
                                                                                            )
                                                                                        )
                                                                                        ( EpaComments
                                                                                            ( []
                                                                                            )
                                                                                        )
                                                                                    )
                                                                                    ( "{abstract:SrcSpan}"
                                                                                    )
                                                                                )
                                                                                ( HsLit
                                                                                    ( EpAnn
                                                                                        ( Anchor
                                                                                            ( "{abstract:RealSrcSpan}"
                                                                                            )
                                                                                            ( UnchangedAnchor
                                                                                            )
                                                                                        )
                                                                                        ( NoEpAnns
                                                                                        )
                                                                                        ( EpaComments
                                                                                            ( []
                                                                                            )
                                                                                        )
                                                                                    )
                                                                                    ( HsString
                                                                                        ( SourceText "\"of\""
                                                                                        )
                                                                                        ( "{abstract:FastString}"
                                                                                        )
                                                                                    )
                                                                                )
                                                                            )
                                                                            ( ( :
                                                                              )
                                                                              ( L
                                                                                  ( SrcSpanAnn
                                                                                      ( EpAnnNotUsed
                                                                                      )
                                                                                      ( "{abstract:SrcSpan}"
                                                                                      )
                                                                                  )
                                                                                  ( HsLit
                                                                                      ( EpAnn
                                                                                          ( Anchor
                                                                                              ( "{abstract:RealSrcSpan}"
                                                                                              )
                                                                                              ( UnchangedAnchor
                                                                                              )
                                                                                          )
                                                                                          ( NoEpAnns
                                                                                          )
                                                                                          ( EpaComments
                                                                                              ( []
                                                                                              )
                                                                                          )
                                                                                      )
                                                                                      ( HsString
                                                                                          ( SourceText "\"evaluation\""
                                                                                          )
                                                                                          ( "{abstract:FastString}"
                                                                                          )
                                                                                      )
                                                                                  )
                                                                              )
                                                                              ( []
                                                                              )
                                                                            )
                                                                          )
                                                                        )
                                                                      )
                                                                    )
                                                                  )
                                                              )
                                                          )
                                                      )
                                                  )
                                                  ( []
                                                  )
                                                )
                                                ( EmptyLocalBinds
                                                    ( NoExtField
                                                    )
                                                )
                                            )
                                        )
                                    )
                                    ( []
                                    )
                                  )
                              )
                              ( FromSource
                              )
                          )
                          ( []
                          )
                      )
                  )
              )
              ( []
              )
            )
          )
          ( Nothing
          )
          ( Nothing
          )
      )
  )
```
</details>

Reply:
Fixed! It will now only parse the toplevel declaration comments.

TIL I learned that I can have a look at GHC AST using `gshow`, I'm super happy. (But the output of gshow is less than perfect).

I lost 6 tests in the battle:

```
    test/Main.hs:37:
    expected: [Just "Evaluate..."]
     but got: [Just "Evaluate...",Just "Refresh..."]
    Use -p '/Produces Evaluate code lenses/' to rerun this test only.
 ```
 
 Let's investigate that later.

Reply:
Well, actually, it is fully solved. The order of the comment is important (i.e. first comments from import, then comments from declarations).